### DISCUSSION TITLE
[Chapter 03] Java의 String은 기본 자료형 (primitive type) 이다?

### DISCUSSION THREAD

**Discussion by Irisation23** [cowaine/is-my-code-that-strange#8] [Chapter 01-03 (설계)] [answered]
p32 최 하단부, p33 최 상부 발췌

```
int와 String처럼 프로그래밍 언어가 표준적으로 제공하는 자료형을 기본 자료형(primitive type)이라고 부릅니다.
```

String이 기본 자료형이라 저자가 표현 했는데 참조형이 아니라 기본 자료형이라 설명한 이유는 `Literal`로 사용할 수 있어서 그런걸까요?
기본 자료형은 8가지로 알고있었는데, 제가 놓친 부분이 있는지 궁금합니다. 

---

> 참조자료
> https://www.baeldung.com/java-primitives

**Comment by firewoody237** [comment, marked as answer]
https://docs.oracle.com/javase%2Ftutorial%2F/java/nutsandbolts/datatypes.html
찾아본 docs를 보면, 자바 언어에서는 아래와 같이 명시되어있습니다.
> A primitive type is predefined by the language and is named by a reserved keyword.
> 
> In addition to the eight primitive data types listed above, the Java programming language also provides special support for character strings via the java.lang.String class.
> 
> The String class is not technically a primitive data type, but considering the special support given to it by the language, you'll probably tend to think of it as such.

`primitive type`은 언어에 의해 이미 만들어져 있고, 8개의 `primitive type`외에 자바는 특별히 character strings을 `java.lang.String`으로 제공한다.

언어 스펙 자체가 "기술적으로 `primitive type`로 볼 순 없으나, 언어에 의해 특별히 지원한다." ---> 첫 줄의 내용과 연결되어 `primitive type`으로 보지않을까 싶습니다..!
(음..근데 그럼 `java.lang.Integer`도 primitive라고 부를 수 있을까하는 의문이..)

**Comment by corock** [comment]
자바 언어에서 원시 자료형(_primitive type_) 은 `boolean`, `char`, `byte`, `short`, `int`, `long`, `float`, `double` 8가지가 맞습니다.
`String` 은 `final` 클래스로 정의되어 있고, 메모리상 스택(_stack_) 영역이 아닌 힙(_heap_) 영역에 저장되는 등 엄연히 참조 자료형입니다.

```java
public final class String
    implements java.io.Serializable, Comparable<String>, CharSequence,
               Constable, ConstantDesc {
    // (...)
}
```

그리고 다른 프로그래밍 언어에서도 문자열 타입은 갖추어야 할 오퍼레이션(_operation_)이 많기 때문에 대부분 원시 자료형으로 다루지 않습니다.

원서를 보진 않았으나 번역에서 오류가 있거나 아니면 저자가 실수로 문장을 작성하지 않았을까 싶습니다.

**Comment by firewoody237** [comment]
코틀린에 뭔가 기존에 정의되어있던 것에 추가 메서드를 더하는 기능이 있는 걸로 알고 있는데..

이건 primitive type에도 사용 가능한가요?

**Reply by coalong** [reply]
노을님: 아마 중위함수일 것 같아요
